### DISCUSSION TITLE
How use getServerSideProps with Postgres?

### DISCUSSION THREAD

**Discussion by jbis9051** [vercel/next.js#12683] [Help]
As per [the Postgres docs](https://node-postgres.com/guides/async-express) I have something similar to this in `./db/index.ts`:

```ts
import {Pool} from 'pg';
const pool = new Pool()
module.exports = {
  query: (text, params) => pool.query(text, params),
}
```

The problem is when I try to use it in a  page.

```ts
// somepage.ts
import pg from "../db";

export async function getServerSideProps() {
    const {row} = await pg.singleRow("SELECT * FROM user WHERE email = $1", ["example@gmail.com"]);
    return row;
}
```

I get the following error

`Module not found: Can't resolve 'dns' in '/path/to/project/node_modules/pg/lib`

This is because the `import pg from "../db"` and therefore `import {Pool} from 'pg';` is running client side and failing. How can I fix this without doing the old method of creating a client side api request for this?

**Comment by timneutkens** [comment]
The code should be tree shaken correctly, you can verify that here:
https://next-code-elimination.now.sh/s/qAvc7UeCM

Are you sure you're exporting this from a page in the pages directory and not somewhere else?

**Reply by jbis9051** [reply]
Hmm. Ok. So here's the situations where its not working. 

**Note**: `../client/hoc/withAuth` imports `../db` and exports two functions. `withAuthServerSideProps`  is the only one that actually uses pg.

The following errors:

```ts
//protected.ts
import {withAuthServerSideProps, withAuthComponent} from '../client/hoc/withAuth';
function protectedPage() {
    return <h1>Hello</h1>
}
export default withAuthComponent(protectedPage);
export const getServerSideProps = withAuthServerSideProps();
```

This doesn't 

```ts
//protected.ts
import {withAuthServerSideProps} from '../client/hoc/withAuth';
function protectedPage() {
    return <h1>Hello</h1>
}

export default protectedPage;
export const getServerSideProps = withAuthServerSideProps();
```

It seems to do this I would need to import two separate files, one with `withAuthServerSideProps` and one with `withAuthComponent`. This is a bit annoying.

**Reply by rafaelalmeidatk** [reply]
What is the code of `withAuth`? You may be doing some side-effects because the code elimination works with two exports from the same module: https://next-code-elimination.now.sh/s/3H8Elw4bl

**Reply by jbis9051** [reply]
@rafaelalmeidatk https://gist.github.com/jbis9051/52eb5c8a015ec44aaff0910cc1e72503

**Reply by Timer** [reply]
You cannot mix server and client code in files that are not the `pages/` directory.

**Reply by jbis9051** [reply]
@Timer So I would be forced to store the functions in two different files? Could next be modified to allow code elimination to files not in pages dir, or would that be technologically very difficult?

Just a bit annoying to have to store two related functions in two different files.
